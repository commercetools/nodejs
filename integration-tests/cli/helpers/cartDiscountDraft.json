{
  "name": {
    "en": "integration-test-cart-discount"
  },
  "value": {
    "type": "relative",
    "permyriad": 1000
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "target": {
    "type": "lineItems",
    "predicate": "1 = 1"
  },
  "sortOrder": "0.1",
  "requiresDiscountCode": true
}
