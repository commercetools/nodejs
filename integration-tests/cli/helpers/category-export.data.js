export default [
  {
    key: 'fooKey',
    name: {
      en: 'categoryA',
    },
    slug: {
      en: 'category-a',
    },
    description: {
      en: 'descA',
    },
    externalId: 'a',
    assets: [],
  },
  {
    key: 'barKey',
    name: {
      en: 'categoryB',
    },
    slug: {
      en: 'category-b',
    },
    description: {
      en: 'descB',
    },
    externalId: 'b',
    assets: [],
  },
  {
    key: 'fooBarKey',
    name: {
      en: 'categoryC',
    },
    slug: {
      en: 'category-c',
    },
    description: {
      en: 'descC',
    },
    externalId: 'c',
    assets: [],
  },
]
