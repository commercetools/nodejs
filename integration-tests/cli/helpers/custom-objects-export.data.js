export default [
  {
    container: 'Ludus',
    key: 'copperKey',
    value: {
      paymentMethod: 'Cash',
      paymentID: '1',
    },
  },
  {
    container: 'Frobozz',
    key: 'jadeKey',
    value: {
      paymentMethod: 'cc',
      paymentID: '2',
    },
  },
  {
    container: 'Syrinx',
    key: 'crystalKey',
    value: {
      paymentMethod: 'new',
      paymentID: '3',
    },
  },
]
