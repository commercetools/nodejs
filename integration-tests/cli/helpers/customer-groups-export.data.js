export default [
  {
    groupName: 'Ludus',
    key: 'copperKey',
  },
  {
    groupName: 'Frobozz',
  },
  {
    groupName: 'Syrinx',
    key: 'crystalKey',
  },
]
