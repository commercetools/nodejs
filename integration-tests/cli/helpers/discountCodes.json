[{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "IT8MKKZX"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITlN85n0"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITDHzsVX"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITlb6Z1u"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITfWgUVd"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITpQgE1P"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITL2pvbZ"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITlrnkj5"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITDJO3LT"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITuIcDI2"
}]
