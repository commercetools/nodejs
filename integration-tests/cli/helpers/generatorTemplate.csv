name.en,name.de,description.en,description.de,cartPredicate,isActive,maxApplications,maxApplicationsPerCustomer
Sammy,Valerian,greatest promo,super angebot,some cart predicate,true,9,4
