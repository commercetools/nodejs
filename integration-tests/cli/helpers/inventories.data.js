export const inventories = [{
  sku: '12345',
  quantityOnStock: 20,
  custom: {
    type: {
      key: 'custom-type',
    },
    fields: {
      description: 'integration tests!! arrgggh',
    },
  },
}]

export const customFields = [{
  key: 'custom-type',
  name: {
    en: 'customized fields',
  },
  description: {
    en: 'custom description',
  },
  resourceTypeIds: ['inventory-entry'],
  fieldDefinitions: [{
    name: 'description',
    type: {
      name: 'String',
    },
    required: false,
    label: {
      en: 'owner',
    },
    inputHint: 'SingleLine',
  }],
}]
