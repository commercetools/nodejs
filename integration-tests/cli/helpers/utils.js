import { createAuthMiddlewareForClientCredentialsFlow } from '@commercetools/sdk-middleware-auth'
import { createClient } from '@commercetools/sdk-client'
import { createRequestBuilder } from '@commercetools/api-request-builder'
import { createHttpMiddleware } from '@commercetools/sdk-middleware-http'
import fetch from 'node-fetch'

export function clearData(apiConfig, entityName) {
  const client = createClient({
    middlewares: [
      createAuthMiddlewareForClientCredentialsFlow({ ...apiConfig, fetch }),
      createHttpMiddleware({ host: apiConfig.apiUrl, fetch }),
    ],
  })

  const service = createRequestBuilder({
    projectKey: apiConfig.projectKey,
  })[entityName]

  const request = {
    uri: service.build(),
    method: 'GET',
  }
  return client.process(request, payload => {
    // Built-in states cannot be deleted
    const results =
      entityName === 'states'
        ? payload.body.results.filter(state => state.builtIn === false)
        : payload.body.results
    return Promise.all(
      results.map(result =>
        client.execute({
          uri: service
            .byId(result.id)
            .withVersion(result.version)
            .build(),
          method: 'DELETE',
        })
      )
    )
  })
}

export function createData(apiConfig, entityName, data, id) {
  const client = createClient({
    middlewares: [
      createAuthMiddlewareForClientCredentialsFlow({ ...apiConfig, fetch }),
      createHttpMiddleware({ host: apiConfig.apiUrl, fetch }),
    ],
  })
  const requestOption = { projectKey: apiConfig.projectKey }
  const service = createRequestBuilder(requestOption)[entityName]
  return Promise.all(
    data.map(_data => {
      if (id) service.byId(id)
      const request = {
        uri: service.build(),
        method: 'POST',
        body: _data,
      }
      return client.execute(request)
    })
  )
}

export function getId(apiConfig, entityName) {
  const client = createClient({
    middlewares: [
      createAuthMiddlewareForClientCredentialsFlow({ ...apiConfig, fetch }),
      createHttpMiddleware({ host: apiConfig.apiUrl, fetch }),
    ],
  })

  const service = createRequestBuilder({
    projectKey: apiConfig.projectKey,
  })[entityName]

  const request = {
    uri: service.build(),
    method: 'GET',
  }

  return client
    .execute(request)
    .then(result => Promise.resolve(result.body.results[0].id))
}
