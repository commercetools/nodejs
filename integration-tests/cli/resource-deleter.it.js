import fetch from 'node-fetch'
import Promise from 'bluebird'
import { createAuthMiddlewareForClientCredentialsFlow } from '@commercetools/sdk-middleware-auth'
import { createClient } from '@commercetools/sdk-client'
import { createRequestBuilder } from '@commercetools/api-request-builder'
import { createHttpMiddleware } from '@commercetools/sdk-middleware-http'
import ResourceDeleter from '@commercetools/resource-deleter'
import { getCredentials } from '@commercetools/get-credentials'
import { exec } from 'mz/child_process'
import { version } from '@commercetools/resource-deleter/package.json'
import * as resources from './helpers/resource-delete.data'
import { createData, clearData } from './helpers/utils'

let projectKey
if (process.env.CI === 'true') projectKey = 'resource-deleter-int-test'
else projectKey = process.env.npm_config_projectkey

describe('Resource Deleter', () => {
  let apiConfig
  let resourceDeleter
  let logger: {
    error: () => {},
    warn: () => {},
    info: () => {},
    debug: () => {},
  }
  const bin = './integration-tests/node_modules/.bin/resource-deleter'

  function getResource(resource) {
    const client = createClient({
      middlewares: [
        createAuthMiddlewareForClientCredentialsFlow({ ...apiConfig, fetch }),
        createHttpMiddleware({ host: apiConfig.apiUrl, fetch }),
      ],
    })

    const service = createRequestBuilder({
      projectKey: apiConfig.projectKey,
    })[resource]

    const request = {
      uri: service.build(),
      method: 'GET',
    }

    return client.execute(request)
  }

  beforeAll(async () => {
    // Get test credentials
    const credentials = await getCredentials(projectKey)
    apiConfig = {
      host: 'https://auth.sphere.io',
      apiUrl: 'https://api.sphere.io',
      projectKey,
      credentials,
    }

    // create resources on API
    await Promise.each(Object.keys(resources), name => {
      return createData(apiConfig, name, resources[name])
    })
  }, 30000)

  // clear resources on API
  afterAll(async () => {
    await Promise.each(Object.keys(resources), name => {
      clearData(apiConfig, name)
    })
  }, 45000)

  describe('CLI basic functionality', () => {
    it('should print usage information given the help flag', async () => {
      const [stdout, stderr] = await exec(`${bin} --help`)
      expect(stderr).toBeFalsy()
      expect(stdout).toMatchSnapshot()
    })

    it('should print the module version given the version flag', async () => {
      const [stdout, stderr] = await exec(`${bin} --version`)
      expect(stderr).toBeFalsy()
      expect(stdout).toBe(`${version}\n`)
    })
  })

  describe.each(Object.keys(resources).reverse())(
    'should delete resource',
    resource => {
      beforeEach(() => {
        const options = {
          apiConfig,
          resource,
          logger,
        }
        resourceDeleter = new ResourceDeleter(options)
      })

      it(`${resource} deleted`, async () => {
        // Check that resource exists
        const payload = await getResource(resource)
        expect(payload.body.results.length).toBeGreaterThanOrEqual(1)

        // Delete resource
        await resourceDeleter.run()

        // Check that resource is deleted
        const newPayload = await getResource(resource)
        expect(newPayload.body.results).toHaveLength(0)
      }, 25000)
    }
  )

  describe('should delete a specific resource', () => {
    const resource = 'channels'
    if (resource === 'channels') {
      beforeEach(() => {
        const options = {
          apiConfig,
          resource,
          logger,
        }
        resourceDeleter = new ResourceDeleter(options)
        return createData(apiConfig, resource, [
          {
            key: 'singleChannel',
            name: {
              en: 'singleChannel',
              de: 'singleChannel',
            },
          },
          {
            key: 'nextChannel',
            name: {
              en: 'nextChannel',
              de: 'nextChannel',
            },
          },
        ])
      })
    }

    it(`The specified ${resource} deleted`, async () => {
      const payload = await getResource(resource)
      expect(payload.body.results.length).toBeGreaterThanOrEqual(1)
      await resourceDeleter.run()
      const newPayload = await getResource(resource)
      expect(newPayload.body.results).toHaveLength(0)
    })
  })

  describe('should delete a specific resource with predicate', () => {
    const resource = 'customerGroups'
    if (resource === 'customerGroups') {
      beforeEach(() => {
        const options = {
          apiConfig,
          resource,
          logger,
          predicate: 'key="CGKey2"',
        }
        resourceDeleter = new ResourceDeleter(options)
        return createData(apiConfig, resource, [
          {
            key: 'CGKey1',
            groupName: 'SampleCGName1',
          },
          {
            key: 'CGKey2',
            groupName: 'SampleCGName2',
          },
          {
            key: 'CGKey3',
            groupName: 'SampleCGName3',
          },
        ])
      })
    }

    it(`The specified ${resource} deleted`, async () => {
      const payload = await getResource(resource)

      // Check the total number of item in the resource.
      expect(payload.body.results).toHaveLength(3)

      // Delete the selected item.
      await resourceDeleter.run()

      // Check that the selected item is deleted
      const newPayload = await getResource(resource)
      expect(newPayload.body.results).toHaveLength(2)
    }, 15000)
  })

  describe('should delete a published product', () => {
    const productType = [
      {
        name: 'sampleProductType',
        key: 'sample-product-123PTKey',
        description: 'Sample Product Type',
        version: 1,
      },
    ]
    const products = [
      {
        key: 'sample-product-123Key',
        name: {
          en: 'sample-product-123',
        },
        slug: {
          en: 'sample-product-123-product-type',
        },
        productType: {
          key: 'sample-product-123PTKey',
          description: 'Sample Product Type',
        },
        version: 1,
        publish: true,
      },
    ]
    const resource = 'products'

    beforeAll(async () => {
      const options = {
        apiConfig,
        resource,
        logger,
      }
      resourceDeleter = new ResourceDeleter(options)
      await createData(apiConfig, 'productTypes', productType)
      await createData(apiConfig, 'products', products)
      return resources
    })
    afterAll(async () => {
      await clearData(apiConfig, 'products')
      return clearData(apiConfig, 'productTypes')
    })

    it(`The published ${resource} deleted`, async () => {
      const payload = await getResource(resource)
      expect(payload.body.results.length).toBe(1)
      await resourceDeleter.run()
      const newPayload = await getResource(resource)
      expect(newPayload.body.results).toHaveLength(0)
    }, 30000)
  })
})
