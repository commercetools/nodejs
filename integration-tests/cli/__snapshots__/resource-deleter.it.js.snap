// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Resource Deleter CLI basic functionality should print usage information given the help flag 1`] = `
"
Usage: resource-deleter [options]
Delete resources from the commercetools platform

Options:
  --help            Show help                                          [boolean]
  --version         Show version number                                [boolean]
  --output, -o      Path to output file.                     [default: \\"stdout\\"]
  --apiUrl          The host URL of the HTTP API service.
                                              [default: \\"https://api.sphere.io\\"]
  --authUrl         The host URL of the OAuth API service.
                                             [default: \\"https://auth.sphere.io\\"]
  --accessToken     CTP client access token.
                    Required scopes: ['manage_products', 'manage_customers',
                    'manage_types']
  --projectKey, -p  API project key.                                  [required]
  --resource, -r    Resource that need to be deleted.                 [required]
  --confirm, -c     Confirm the resource to delete. [boolean] [default: \\"false\\"]
  --where, -w       Specify where predicate.
  --logLevel        Logging level: error, warn, info or debug. [default: \\"info\\"]
  --prettyLogs      Pretty print logs to the terminal                  [boolean]
  --logFile         Path to where to save logs file.
                              [string] [default: \\"resources-deleted-report.log\\"]
"
`;
